### ISSUE TITLE
Old device twin value missing from TwinChangeEvents

### ISSUE THREAD

**Issue by yosephlodamo** [Azure/azure-iot-sdk-node#64]
OS: Windows 10
Node.js: v8.1.2
Npm packages:
    "azure-event-hubs": "0.0.8",
    "azure-iothub": "1.1.12",

I [created a route](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-messages-read-custom) for TwinChangeEvents in Microsoft Azure portal and I can [Receive device-to-cloud messages](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-node-node-getstarted#receive-device-to-cloud-messages) when device twin changes. But the message only contains the new device twin value. I needed the old value for analysis so that I don’t have to analyze the whole devices every time a device twin property changes.

Is there any way I can get the old value?

Code I used to receive the message (https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-node-node-getstarted#receive-device-to-cloud-messages):
```
var EventHubClient = require('azure-event-hubs').Client;
var connectionString = '{iothub connection string}';
var printError = function (err) {
  console.log(err.message);
};

var printMessage = function (message) {
  console.log('Message received: ');
  console.log(JSON.stringify(message.body));
  console.log(JSON.stringify(message.body.properties.reported)); // New value
  console.log('');
};
var client = EventHubClient.fromConnectionString(connectionString);
client.open()
    .then(client.getPartitionIds.bind(client))
    .then(function (partitionIds) {
        return partitionIds.map(function (partitionId) {
            return client.createReceiver('$Default', partitionId, { 'startAfterTime' : Date.now()}).then(function(receiver) {
                console.log('Created partition receiver: ' + partitionId)
                receiver.on('errorReceived', printError);
                receiver.on('message', printMessage);
            });
        });
    })
    .catch(printError);
```

**Comment by ailn** [comment]
Hello @yosephlodamo,

Current implementation of Twin change events does not contain changed properties' old value. You can only get 'PATCH' (what was applied to prev state). 
Moreover, there's no guarantee that you'll get all the corresponding events: it is possible that you receive the whole twin state (in case opType='replaceTwin').

In other words we do not suport auditing at the moment. 
Primary usecase of twin change events is data store synchronization.

**Comment by romankiss** [comment]
Hi @ailn,

Beside that, it looks like, the current implementation of TwinChangeEvents has a bug when the event is published. Now, it is working like TwinUpdateEvents, because the event is generated when this service is invoked, not only for the cases when the device twin state has been changed such as new state.

In other words, the event is generated every time when the twin update has been called doesn't matter if the patch is empty or the same value and the result is a new device twin version number. 

I think, the TwinChangeEvents should be have ONLY events when the device twin state has been changed.

**Comment by ailn** [comment]
Hi @yosephlodamo,

To clarify couple things:
1. When you do PATCH of /twins/{id} resource and values are same we DO increase version of the twin resource and this is by design. That is why we also send corresponding notification.
2. When PATCH is empty we DO NOT update twin resource and do not send notification.

If (2.) is not what you get then please check whether twins/{id} resource has changed or not after empty PATCH. If it has changed then please post an example of the PATCH payload you're sending.

**Comment by romankiss** [comment]
Hi @ailn,

The number of twins version should be indicated a new device state, otherwise it is useless. 
Based on your case 1, the changing of the version number is a fuzzy value and it is not suitable for query key in the persisted device twins changes.

If the case 1 is a valid design requirement, than the case 2 is irrelevant.

Anyway, the following are few examples for the case 2, when the version number is still incremented and notification is sent to the TwinChangeEvents Routes.

- MQTT Device (_$iothub/twin/PATCH/properties/reported/?$rid=1_)
payloads:
{}
{ "abc": {} }
{ "abc": null }
 
- Service-facing endpoint ( _/twins/{id}_ )
payloads:
{
  "properties": {
    "desired": null
  }
}
{
  "properties": null
}
{
  "tags":null
}

Still I think, that the version number should be represented a number of changes in the device twins. In other words, it should be here applied an observer event pattern such as notifications of any twins' state changes. Note, that the state of the device twins are tags, properties.desired and properties.reported (not the version). 
It is a good patterns & practices in the event driven model to notify its changes especial for very large model with hundred thousands devices.

Besides that, I think, the case 1 will create some trouble for implementation of the delta state (similarly to the AWS), where a notification is sent for synchronizing of the desired and reported states.

Thanks
Roman

**Comment by ailn** [comment]
Hi @romankiss,

We treat _null_ as 'delete property'. So when you send _null_ we treat is as not empty payload. I beleive this should be stressed in documentation. 

For all your Service-facing examples we treat payload as not empty. In fact you cannot send a PATCH with an empty payload as you'll get a 400 in response saing "Both tags and desired properties cannot be null".

For MQTT examples the payload is not 'empty' (Maybe empty is not the right word here - not null) as well.

**Comment by ailn** [comment]
@romankiss, regarding incrementing version even if values not changed... PATCH is not intended to be sent untill you need to change something. So we treat a PATCH with unchanged values as intentional and we track this in metadata.

**Comment by romankiss** [comment]
Hi @ailn,

Thanks for your quick reply. My point is the same, such as based on the observer event patterns & practices, the notification should be generated only in the case when the state has been changed. 

In other hands, when we have more notifications for cases like a PATCH with the same value, DELETE not exist property, etc. it will cost an additional call for GetDeviceTwin. 

The source _TwinChangeEvents_ should be represented source of events to notify that the device twin state has been changed. The same, if the MQTT Device subscribed for topic _$iothub/twin/PATCH/properties/desired/#_ it will expect some changes in the desired properties and the reason to call (billing call) the backend for device twin state. 

Note, that the service-facing endpoint PATCH is working well for desired properties, for example::

{
  "properties": {
    "desired": {
      "abcd": null
    }
  }
}

{
  "properties": {
    "desired": {
      "abcd": {}
    }
  }
}

There is no extra notifications when the PATCH is called multiple times with the above payload. Only problem is still here when the PATCH is used for the same value multiple times.

The situation on the device-facing endpoint is completely different. For example, the payload sent by MQTTBox Client (where the "abcd" is not exist property):

{
   "abcd":null
}

will generate a notification to the _TwinChangeEvents_ every time when the device published it:
{
  "properties": {
    "reported": {
      "abcd": null,
      "$metadata": {
        "$lastUpdated": "2017-07-30T12:33:33.1049809Z",
        "abcd": null
      },
      "$version": 215
    }
  }
}

the same situation is for payload:
{
   "abcd":{}
}


In other words, the behavior between the desired and reported properties are different at the TwinChangeEvents source. 

Thanks
Roman

**Comment by ailn** [comment]
Hi Roman,

Sorry for the late response. I confirm there's a bug for reported properties in case a device sends non-existent property set to null: it will NOT add anything to the state but WILL send notification. 

Thanks for finding this,
Andrei

**Comment by romankiss** [comment]
Hi @ailn,

The _TwinChangeEvents_ notifications pattern should be the same from desired and reported properties of the device twin.. 

The behavior of the notification from the desired properties for non-existent property set to null is: **NOT added anything to the state and NOT sent any notification**. This is a correct implemented observer event pattern and it should be applied also at the reported properties.

Thanks
Roman

**Comment by pierreca** [comment]
Update: we have a fix for this coming in the service that will require the SDK to latch onto a new api-version string. it's reasonable to expect this fix to be available sometime in February.

**Comment by pierreca** [comment]
@romankiss the fix has been deployed by the service (some time ago) and we finally got around updating the SDK API version, so this should be fixed with azure-iothub@1.2.5

@ailn Thanks for your help!
